### PR TITLE
Approach to assignment 2

### PR DESCRIPTION
**Pull request by katermar**
both questions implemented in methods in BinaryTree class

### CODE REVIEW

File: BinaryTree.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
+package com.katermar;
+
+/**
+ * Created by USER on 5/23/2017.
```
Comment (elifguler96) [inline BinaryTree.java]: You could just delete this one I guess :)

File: BinaryTree.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    /*Inner class for Binary Tree node
+    * @field value stores integer value
+    * @field left, right store left and right subtrees*/
+    public static class TreeNode {
+        private int value;
+        private TreeNode left, right, nextRight;
+
+        TreeNode(int item) {
+            value = item;
+            left = right = nextRight = null;
```
Comment (elifguler96) [inline BinaryTree.java]: In my opinion, writing these seperately look nicer.

File: BinaryTree.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+ */
+public class BinaryTree {
+    TreeNode root;
+
+    /*Inner class for Binary Tree node
+    * @field value stores integer value
+    * @field left, right store left and right subtrees*/
+    public static class TreeNode {
+        private int value;
+        private TreeNode left, right, nextRight;
```
Comment (elifguler96) [inline BinaryTree.java]: What do you use nextRight for?

File: BinaryTree.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    /*@param value - value to insert*/
+    public void insert(int value) {
+        root = insert(root, value);
+    }
+
+    /*Inserts value to the tree
+    * @param node - root
+    * @param value - value to insert
+    * @return root*/
+    private static TreeNode insert(TreeNode node, int value) {
```
Comment (elifguler96) [inline BinaryTree.java]: With this insert you can only generate binary _search_ trees, but you may want to test your code with regular binary trees. So, if you want maybe you could change this insert a bit.

File: BinaryTree.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        }
+        return node;
+    }
+
+    /*Finds lowest common ancestor of two nodes
+    * @param root
+    * @param a - first node value
+    * @param b - second node value
+    * @return lowest common ancestor if both nodes are present
+    * int the tree, if one is not present - returns existing node, as LCA
```
Comment (elifguler96) [inline BinaryTree.java]: Think there is a typo here 😄 int -> in

File: CompareStringPermutation.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Created by USER on 5/15/2017.
+ */
+public class CompareStringPermutation {
+
+    public boolean compare(String first, String last) {
+        char[] a = first.toLowerCase().toCharArray();
+        char[] b = last.toLowerCase().toCharArray();
+
+        Arrays.sort(a);
```
Comment (elifguler96) [inline CompareStringPermutation.java]: As discussed during the Hangout, time complexity of sort is O(NlogN) and there is a solution which works in O(N). Just wanted to point this out :)

File: CompareStringPermutation.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.katermar;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+/**
+ * Created by USER on 5/15/2017.
```
Comment (elifguler96) [inline CompareStringPermutation.java]: Again, you could delete this

File: KthLastElement.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
+package com.katermar;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Created by USER on 5/15/2017.
```
Comment (elifguler96) [inline KthLastElement.java]: This one too :D

File: Node.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
+package com.katermar;
+
+/**
+ * Created by USER on 5/17/2017.
```
Comment (elifguler96) [inline Node.java]: And this

File: BinaryTree.java (hunk 4 above)
Comment (katermar) [inline BinaryTree.java]: Can you pls explain me why it makes sence in my solution? Cuz I don't use the fact, that it is the binary search tree. 
I populate my tree by array, so I can just change their order to change height of the tree, but value in the node may not bother us.

Comment (elifguler96) [inline BinaryTree.java]: Yeah I know that you didn't solve it assuming it will be a binary _search_ tree, but someone asked this question on Slack and this is the reply he got:
image
Just wanted to point this out :)

Comment (katermar) [inline BinaryTree.java]: yeap, I've seen it :)
but don't _you_ think that **insert()** logic looks nicer and maybe is easier to test and to understand?

Comment (elifguler96) [inline BinaryTree.java]: Yes it does look nicer like this and it will probably make no difference, but I kinda agree that we should test each case.

File: BinaryTree.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            return value;
+        }
+
+        public TreeNode getLeft() {
+            return left;
+        }
+
+        @Override
+        public String toString() {
+            return "TreeNode{" +
```
Comment (brzozan) [inline BinaryTree.java]: Alternatively you could use String.format, which might be a bit less efficient, but results in a nicer code.

File: BinaryTree.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+ * Created by katermar on 5/23/2017.
+ */
+public class BinaryTree {
+    TreeNode root;
+
+    /*Inner class for Binary Tree node
+    * @field value stores integer value
+    * @field left, right store left and right subtrees*/
+    public static class TreeNode {
+        private int value;
```
Comment (brzozan) [inline BinaryTree.java]: value field could be made final, as it is assigned only in constructor.

File: BinaryTree.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    * @param root
+    * @param a - first node value
+    * @param b - second node value
+    * @return lowest common ancestor if both nodes are present
+    * in the tree, if one is not present - returns existing node, as LCA
+    * if both don't exist - returns null
+    * if a key is ancestor of other, then the ancestor key becomes LCA*/
+    public TreeNode findLowestCommonAncestor(TreeNode root, int a, int b) {
+        // Base case
+        if (root == null) return null;
```
Comment (brzozan) [inline BinaryTree.java]: Please try to avoid both single-line if statements and ones without {}. Always including brackets does not increase code size that much, but highly reduces mental power needed to read code, and thus makes it easier to avoid silly mistakes while modifying it in the future.

File: BinaryTree.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                /   \
+               1     4
+                    /  \
+                   3    5
+                         \
+                          7
+        */
+        int[] array = {2, 1, 4, 3, 5, 7};
+        tree.populateTree(array);
+        tree.preorderTraversal(tree.root);
```
Comment (brzozan) [inline BinaryTree.java]: This is not really necessary. It is good to keep code to bare minimum to avoid introducing noise.

File: BinaryTree.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+            return right;
+        }
+
+    }
+
+    /* Function to print all ancestors of purposed value
+    * @param node - root node
+    * @param value - value of the node on which we want to print ancestors
+    * @return true, if value exists in this tree, otherwise - false*/
+    boolean printAllAncestors(TreeNode node, int value)
```
Comment (brzozan) [inline BinaryTree.java]: What if we at some point we need to do something useful (rather than just print) with node's ancestors? Or test it more extensively in automated fashion?

Could you please refactor code so that finding them is separate from printing them?

File: BinaryTree.java [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        int[] array = {2, 1, 4, 3, 5, 7};
+        tree.populateTree(array);
+        tree.preorderTraversal(tree.root);
+        System.out.println();
+
+        if (!tree.printAllAncestors(tree.root, 4))
+            System.out.println("Element doesn't exist");
+
+        System.out.println();
+        System.out.println(tree.findLowestCommonAncestor(tree.root, 4, 10));
```
Comment (brzozan) [inline BinaryTree.java]: Finding ancestors, and especially common ancestors of two (or more) nodes, has a few corner cases that are not covered by this println. Plus it relies on human reading output of the file and verifying if it is correct or not. It makes it easy for bugs to slip through.

Could you please create tests covering a few such scenarios? You could either use [JUnit](https://github.com/junit-team/junit4/wiki/Getting-started) or go with your own methods that would compare result of the methods with expected one and output status that would indicate if each test passed or not?

File: BinaryTree.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
+package com.katermar;
```
Comment (brzozan) [inline BinaryTree.java]: You might consider creating subpackages for each assignment to clearly separate code for each of them and to avoid naming conflicts (BinaryTree.TreeNode vs TreeNode from previous assignment).

Comment (katermar) [inline BinaryTree.java]: speaking about different branches for each assignment. can i just have single branch for current assignment? and to merge previous assignments to the master branch after reviews?

Comment (brzozan) [inline BinaryTree.java]: It is OK to leave assignment 2 on current branch and merge it to master once review is finished. Moving it to a new branch would be a bit problematic and would probably lose review comments.
But for assignment 3 (and 4 and so on) please create new branches. Thanks!

File: BinaryTree.java [hunk 17]
```diff
@@ -153,13 +166,33 @@ public static void main(String args[])
...
+        } else {
+            System.out.println("Ancestors - " + ancestorList);
+        }
 
         System.out.println();
-        System.out.println(tree.findLowestCommonAncestor(tree.root, 4, 10));
+        System.out.print(tree.findLCAtest(7, 3));
+    }
+
+    public String findLCAtest(int a, int b) {
```
Comment (brzozan) [inline BinaryTree.java]: Test methods are typically not part of the class that is being tested. It prevents polluting class' interface and lets one avoid relying on some internal data for testing.

In this case better design would be to have class BinaryTree that only implements binary tree, and another class (Main, BinaryTreeTest or similar) that would contain main() and any test code needed.

File: BinaryTree.java [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+ */
+public class BinaryTree {
+    TreeNode root;
+
+    /*Inner class for Binary Tree node
+    * @field value stores integer value
+    * @field left, right store left and right subtrees*/
+    public static class TreeNode {
+        private final int value;
+        private TreeNode left, right;
```
Comment (brzozan) [inline BinaryTree.java]: Declaring multiple fields or variables in a single line is more difficult to read than a single declaration per line. Could you please change code to split declarations of left and right?

The same applies to code from assignment 1, in which I missed it. You could change it there as well or leave as is.

File: BinaryTree.java [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+
+        TreeNode(int item) {
+            value = item;
+            left = right = null;
```
Comment (brzozan) [inline BinaryTree.java]: Just like multiple declarations in a single line, such chain assignments are more difficult to read and more error prone. Please avoid them.

Also in this case initialization with null is not really needed - fields have null value by default, but feel free to keep explicit initialization (with single assignment per line) if you like it. It does not hurt to be explicit.

File: BinaryTree.java [hunk 20]
```diff
@@ -0,0 +1,198 @@
+package com.katermar.assignment2;
```
Comment (brzozan) [inline BinaryTree.java]: Classes that belong to different subpackages usually live in separate subfolders, so that you can have two classes with the same name, each in diffrent package.

Proper path for BinaryTree.java would be com/katermar/assignment2/BinaryTree.java. Could you please move all .java files to proper subfolders? It is quite [easy](https://help.github.com/articles/moving-a-file-to-a-new-location/).

File: BinaryTree.java [hunk 21]
```diff
@@ -150,49 +150,49 @@ public TreeNode findLowestCommonAncestor(TreeNode root, int a, int b) {
...
-            return LCA + " - One of the nodes is LCA";
-        }
-        if (LCA.getValue() == this.root.getValue()) {
-            return LCA + " - LCA is a root";
-        }
-        else {
-            return LCA + " - is LCA";
-        }
-    }
+//    /* Driver program to test above functions */
```
Comment (brzozan) [inline BinaryTree.java]: Please delete commented out code.

File: BinaryTreeTest.java [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+            System.out.println("Ancestors - " + ancestorList);
+        }
+
+        System.out.println();
+        checkLCA(4, tree, 7, 3);
+        checkLCA(2, tree, 2, 1);
+        checkLCA(10, tree, 2, 1);
+    }
+
+    public static String findLCAtest(BinaryTree tree, int a, int b) {
```
Comment (brzozan) [inline BinaryTreeTest.java]: It is not really needed any more - the whole check happens in checkLCA. Of course you can keep this method if you wish, but please remove call to it in checkLCA.

File: BinaryTreeTest.java [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+                   3    5
+                         \
+                          7
+        */
+        int[] array = {2, 1, 4, 3, 5, 7};
+        tree.populateTree(array);
+        System.out.println();
+
+        List<BinaryTree.TreeNode> ancestorList = new LinkedList<>();
+        if (!tree.findAllAncestors(ancestorList, tree.root, 7)) {
```
Comment (brzozan) [inline BinaryTreeTest.java]: Could you add tests for findAllAncestors too?

File: assignment2/BinaryTree.java [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+            return true;
+        }
+
+        //else return false
+        return false;
+    }
+
+    /*Traverse tree in preorder
+    * @param root - tree root*/
+    public void preorderTraversal (TreeNode root){ //метод для обхода в ширину
```
Comment (brzozan) [inline assignment2/BinaryTree.java]: Minor comment regarding design: preorderTraversal is a method of BinaryTree and I would expect it to operate on the whole tree. As such the root parameter is not needed and makes method more difficult to use.

Better design would be BinaryRoot.preorderTraversal() without parameters calling TreeNode.preorderTraversal() (also without parameters - in such case root == this), which then descends recursively.

File: assignment2/BinaryTree.java [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    /*Populates tree with the array in order
+    * @param array - array which will fill the tree*/
+    public boolean populateTree(int[] array) {
+        for (int i = 0; i < array.length; i++) {
+            this.insert(array[i]);
+        }
+
+        return true;
```
Comment (brzozan) [inline assignment2/BinaryTree.java]: Since this method always returns true, it does not really make sense to have boolean as return type. Just void would be enough.

File: assignment2/BinaryTree.java [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+
+    /*Finds lowest common ancestor of two nodes
+    * @param root
+    * @param a - first node value
+    * @param b - second node value
+    * @return lowest common ancestor if both nodes are present
+    * in the tree, if one is not present - returns existing node, as LCA
+    * if both don't exist - returns null
+    * if a key is ancestor of other, then the ancestor key becomes LCA*/
+    public TreeNode findLowestCommonAncestor(TreeNode root, int a, int b) {
```
Comment (brzozan) [inline assignment2/BinaryTree.java]: Again, requiring user of the method to specify starting node makes it more difficult to use. Could you remove root parameter from public method? Then you can either have private helper method that does use root and is used for recursion (just like you did for insert - good job!), or you could move implementation to TreeNode.

File: assignment2/BinaryTreeTest.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        if (LCA.getValue() == tree.root.getValue()) {
+            return LCA + " - LCA is a root";
+        }
+        else {
+            return LCA + " - is LCA";
+        }
+    }
+
+    public static void checkLCA(int expectedResult, BinaryTree tree, int a, int b) {
```
Comment (brzozan) [inline assignment2/BinaryTreeTest.java]: No need to make it public. Private would be enough.

**Comment by brzozan** [comment]
Hi Kate,

I see that you have merged the branch before I had a chance to do final review of the changes. When doing the next assignments please wait for approval. Thanks!

**Comment by katermar** [comment]
Hi, 
Sorry. Next time I should wait until you'll approve the latest version?

**Comment by brzozan** [comment]
Yes. Usually the process is that one waits for approval from all reviewers before merging. Sometimes one approval is sufficient, but that is more of an exception rather than a rule.
